From 21f7f5ef251d25d8857cfd8a0b547bea8ec79d55 Mon Sep 17 00:00:00 2001
From: Jialin Zhang <zhangjialin11@huawei.com>
Date: Wed, 6 Jul 2022 12:24:12 +0800
Subject: [PATCH 3302/4585] [Huawei] arm64/mpam: update last_cmd_status in
 parse_cache() and parse_bw()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187154

--------------------------------

Update last_cmd_status to tell the reasons of returning 'Invalid argument'
in parse_cache() and parse_bw().

Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 arch/arm64/kernel/mpam/mpam_resctrl.c | 26 ++++++++++++++++++++------
 1 file changed, 20 insertions(+), 6 deletions(-)

diff --git a/arch/arm64/kernel/mpam/mpam_resctrl.c b/arch/arm64/kernel/mpam/mpam_resctrl.c
index f9a360a4c718..7400db4ab9ea 100644
--- a/arch/arm64/kernel/mpam/mpam_resctrl.c
+++ b/arch/arm64/kernel/mpam/mpam_resctrl.c
@@ -310,11 +310,15 @@ parse_cache(char *buf, struct resctrl_resource *r,
 		return -EINVAL;
 	}
 
-	if (kstrtoul(buf, rr->ctrl_features[type].base, &data))
+	if (kstrtoul(buf, rr->ctrl_features[type].base, &data)) {
+		rdt_last_cmd_printf("Non-hex character in the mask %s\n", buf);
 		return -EINVAL;
+	}
 
-	if (data >= rr->ctrl_features[type].max_wd)
+	if (data >= rr->ctrl_features[type].max_wd) {
+		rdt_last_cmd_puts("Mask out of range\n");
 		return -EINVAL;
+	}
 
 	cfg->new_ctrl[type] = data;
 	cfg->have_new_ctrl = true;
@@ -339,20 +343,30 @@ parse_bw(char *buf, struct resctrl_resource *r,
 	case QOS_MBA_MAX_EVENT_ID:
 	case QOS_MBA_PBM_EVENT_ID:
 	case QOS_MBA_MIN_EVENT_ID:
-		if (kstrtoul(buf, rr->ctrl_features[type].base, &data))
+		if (kstrtoul(buf, rr->ctrl_features[type].base, &data)) {
+			rdt_last_cmd_printf("Non-decimal digit in MB value %s\n", buf);
 			return -EINVAL;
-		if (data < r->mbw.min_bw)
+		}
+		if (data < r->mbw.min_bw) {
+			rdt_last_cmd_printf("MB value %ld out of range [%d,%d]\n", data,
+					r->mbw.min_bw, rr->ctrl_features[type].max_wd - 1);
 			return -EINVAL;
+		}
 		data = roundup(data, r->mbw.bw_gran);
 		break;
 	default:
-		if (kstrtoul(buf, rr->ctrl_features[type].base, &data))
+		if (kstrtoul(buf, rr->ctrl_features[type].base, &data)) {
+			rdt_last_cmd_printf("Non-decimal digit in MB value %s\n", buf);
 			return -EINVAL;
+		}
 		break;
 	}
 
-	if (data >= rr->ctrl_features[type].max_wd)
+	if (data >= rr->ctrl_features[type].max_wd) {
+		rdt_last_cmd_printf("MB value %ld out of range [%d,%d]\n", data,
+				r->mbw.min_bw, rr->ctrl_features[type].max_wd - 1);
 		return -EINVAL;
+	}
 
 	cfg->new_ctrl[type] = data;
 	cfg->have_new_ctrl = true;
-- 
2.17.1

